src/panels/KaitoModelsPanel.ts

```diff
@@ -270,46 +264,57 @@ export class KaitoModelsPanelDataProvider implements PanelDataProvider<"kaitoMod
             } else {
                 quotaAvailable = true;
             }
-
-            this.checkingGPU = false;
             if (!quotaAvailable) {
+                this.checksInProgress = false;
                 return;
             }
-            const tempFilePath = join(tmpdir(), `kaito-deployment-${Date.now()}.yaml`);
-            writeFileSync(tempFilePath, yaml, "utf8");
-            const command = `apply -f ${tempFilePath}`;
-            const kubectlresult = await invokeKubectlCommand(this.kubectl, this.kubeConfigFilePath, command);
-            if (failed(kubectlresult)) {
-                vscode.window.showErrorMessage(`Error during deployment: ${kubectlresult.error}`);
+
+            // Deployment cancellation check
+            if (this.cancelTokens.get(model)) {
+                this.handleDeploymentCancellation(model);
+                return;
+            }
+            // Deploying model
+            const deploymentResult = await deployModel(yaml, this.kubectl, this.kubeConfigFilePath);
+            if (failed(deploymentResult)) {
+                vscode.window.showErrorMessage(`Error deploying 'workspace-${model}': ${deploymentResult.error}`);
+                this.checksInProgress = false;
                 return;
             }
+            // Final checksInProgress set to false to allow for future deployments
+            this.checksInProgress = false;
         } catch (error) {
-            this.checkingGPU = false;
+            this.checksInProgress = false;
             vscode.window.showErrorMessage(`Error during deployment: ${error}`);
         }
-        this.cancelTokens.set(model, false);
-        this.handleUpdateStateRequest(model, webview);
+        // end state updates if user cancelled out of view
+        if (this.cancelTokens.get(model)) {
+            vscode.window.showErrorMessage(`Deployment cancelled for ${model}`);
+            return;
+        }
+        // if cancel token is set for any other model, just return maybe?
+        this.updateProgress(model, webview);
         let progress = await this.getProgress(model);
-        while (!this.nullIsFalse(progress.workspaceReady) && !this.cancelTokens.get(model)) {
-            await this.handleUpdateStateRequest(model, webview);
-            progress = await this.getProgress(model);
+
+        // Continuously polls deployment progress until completion, cancellation, or panel disposal.
+        while (
```

Comment:
💡 Please  Note: please be careful, because usage of `while` and the potential risks of while loops, such as infinite loops caused by improper termination conditions, I mention because you might know this whole logic flow better than me. =)